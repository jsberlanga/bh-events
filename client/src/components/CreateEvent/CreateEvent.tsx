import { ErrorMessage } from 'components/ErrorMessage';
import { useCreateForm, useEvents } from 'hooks';
import * as React from 'react';
import styled from 'styled-components';

const Form = styled.form`
  display: flex;
  flex-direction: column;
  width: 15rem;
  margin-bottom: 1rem;
  padding: 14px;
`;

const Label = styled.label`
  font-size: 0.8rem;
  margin-bottom: 0.7rem;
`;

const InputField = styled.input`
  box-sizing: border-box;
  width: 100%;
  padding: 6px;
  font-size: 1rem;
`;

const CreateEvent = () => {
  const { events } = useEvents();
  const { event, actionState, handleChange, handleSubmit } = useCreateForm(
    events,
  );

  return (
    <React.Fragment>
      <h2>Create a new event</h2>
      <Form onSubmit={handleSubmit} className="withBorder">
        <Label htmlFor="name">
          Name
          <InputField
            className="withBorder"
            id="name"
            type="text"
            required
            value={event.name}
            onChange={(e) => handleChange(e, 'name')}
          />
        </Label>
        <Label htmlFor="lastname">
          Lastname
          <InputField
            className="withBorder"
            id="lastname"
            type="text"
            required
            value={event.lastname}
            onChange={(e) => handleChange(e, 'lastname')}
          />
        </Label>
        <Label htmlFor="email">
          Email
          <InputField
            className="withBorder"
            id="email"
            type="email"
            required
            value={event.email}
            onChange={(e) => handleChange(e, 'email')}
          />
        </Label>
        <Label htmlFor="date">
          Date
          <InputField
            className="withBorder"
            id="date"
            type="date"
            required
            value={event.date}
            onChange={(e) => handleChange(e, 'date')}
          />
        </Label>
        <InputField
          type="submit"
          value={`creat${actionState.loading ? 'ing' : 'e'}`}
          className="btn btn--submit"
        />
      </Form>
      {actionState.error ? (
        <ErrorMessage>{actionState.error}</ErrorMessage>
      ) : null}
    </React.Fragment>
  );
};

export default CreateEvent;
