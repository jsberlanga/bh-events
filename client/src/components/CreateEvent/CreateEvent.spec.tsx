import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CreateEvent from './CreateEvent';

const mockEvent = {
  name: 'test_name',
  lastname: 'test_lastname',
  email: 'test@example.com',
  date: '2020/01/01',
};

describe('CreateEvent', () => {
  it('should render with styles', () => {
    const { container } = render(<CreateEvent />);

    expect(container).toMatchSnapshot();
  });

  it('should render form correctly', () => {
    render(<CreateEvent />);

    expect(screen.getByLabelText('Name')).toBeInTheDocument();
    expect(screen.getByLabelText('Lastname')).toBeInTheDocument();
    expect(screen.getByLabelText('Email')).toBeInTheDocument();
    expect(screen.getByLabelText('Date')).toBeInTheDocument();
  });

  it('should display an error message if form data is incomplete', async () => {
    render(<CreateEvent />);

    const createButton = screen.getByRole('button');

    userEvent.click(createButton);

    await waitFor(() => {
      expect(
        screen.getByText('An error ocurred: Conflict'),
      ).toBeInTheDocument();
    });
  });

  it('should handle change event', async () => {
    render(<CreateEvent />);

    userEvent.type(screen.getByLabelText('Name'), mockEvent.name);
    userEvent.type(screen.getByLabelText('Lastname'), mockEvent.lastname);
    userEvent.type(screen.getByLabelText('Email'), mockEvent.email);

    expect(screen.getByLabelText('Name').getAttribute('value')).toBe(
      mockEvent.name,
    );
    expect(screen.getByLabelText('Lastname').getAttribute('value')).toBe(
      mockEvent.lastname,
    );
    expect(screen.getByLabelText('Email').getAttribute('value')).toBe(
      mockEvent.email,
    );
  });

  it('should display an in progress indication when submitting the form', async () => {
    render(<CreateEvent />);

    userEvent.type(screen.getByLabelText('Name'), mockEvent.name);
    userEvent.type(screen.getByLabelText('Lastname'), mockEvent.lastname);
    userEvent.type(screen.getByLabelText('Email'), mockEvent.email);

    expect(screen.getByLabelText('Name').getAttribute('value')).toBe(
      mockEvent.name,
    );
    expect(screen.getByLabelText('Lastname').getAttribute('value')).toBe(
      mockEvent.lastname,
    );
    expect(screen.getByLabelText('Email').getAttribute('value')).toBe(
      mockEvent.email,
    );
    const createButton = screen.getByRole('button');

    userEvent.click(createButton);

    expect(screen.getByRole('button').getAttribute('value')).toBe('creating');

    await waitFor(() => {
      expect(screen.getByRole('button').getAttribute('value')).toBe('create');
    });

    expect(screen.getByLabelText('Name').getAttribute('value')).toBe('');
    expect(screen.getByLabelText('Lastname').getAttribute('value')).toBe('');
    expect(screen.getByLabelText('Email').getAttribute('value')).toBe('');
  });
});
