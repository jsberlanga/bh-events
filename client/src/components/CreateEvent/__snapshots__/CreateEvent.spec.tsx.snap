// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEvent should render with styles 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 15rem;
  margin-bottom: 1rem;
  padding: 14px;
}

.c1 {
  font-size: 0.8rem;
  margin-bottom: 0.7rem;
}

.c2 {
  box-sizing: border-box;
  width: 100%;
  padding: 6px;
  font-size: 1rem;
}

<div>
  <h2>
    Create a new event
  </h2>
  <form
    class="c0 withBorder"
  >
    <label
      class="c1"
      for="name"
    >
      Name
      <input
        class="c2 withBorder"
        id="name"
        required=""
        type="text"
        value=""
      />
    </label>
    <label
      class="c1"
      for="lastname"
    >
      Lastname
      <input
        class="c2 withBorder"
        id="lastname"
        required=""
        type="text"
        value=""
      />
    </label>
    <label
      class="c1"
      for="email"
    >
      Email
      <input
        class="c2 withBorder"
        id="email"
        required=""
        type="email"
        value=""
      />
    </label>
    <label
      class="c1"
      for="date"
    >
      Date
      <input
        class="c2 withBorder"
        id="date"
        required=""
        type="date"
        value="2021-01-22"
      />
    </label>
    <input
      class="c2 btn btn--submit"
      type="submit"
      value="create"
    />
  </form>
</div>
`;
